### PR TITLE
Rewrite OrderPdfs to support multiple orders at once

### PR DESCRIPTION
**Pull request by paroga**
This implements `MultipleOrdersBy*` from the `current_orders` in a clean way, so we can implement #527 in a clean way.

### CODE REVIEW

File: app/models/group_order_article.rb [hunk 1]
```diff
@@ -195,4 +195,8 @@ def total_price(order_article = self.order_article)
   def result_manually_changed?
     result != result_computed unless result.nil?
   end
+
+  def quantity_with_tolerance
+    tolerance > 0 ? "#{quantity} + #{tolerance}" : "#{quantity}"
+  end
```
Comment (wvengen) [inline app/models/group_order_article.rb]: Are you sure to put this here? In html we may want to include coloring, for example. Maybe keep it in a pdf-related module?

File: app/documents/order_by_articles.rb [hunk 2]
```diff
@@ -18,59 +12,30 @@ def title
...
-
-      rows = []
       dimrows = []
-      each_group_order_article_for(order_article) do |goa|
+      rows = [[
+        GroupOrder.human_attribute_name(:ordergroup),
+        GroupOrderArticle.human_attribute_name(:ordered),
+        GroupOrderArticle.human_attribute_name(:received),
+        GroupOrderArticle.human_attribute_name(:total_price)
+      ]]
```
Comment (wvengen) [inline app/documents/order_by_articles.rb]: Yeah, much better to maintain :)

File: lib/order_pdf.rb [hunk 3]
```diff
@@ -1,104 +1,130 @@
...
+
+      results = goa_records.group_by(&:first).transform_values do |value|
+        grouped_value = value.group_by(&:second)
+        group_ids.map do |group_id|
+          grouped_value[group_id].try(:first).try(:third)
+        end
+      end
+
+      yield go_records, results
+      offset += batch_size
```
Comment (wvengen) [inline lib/order_pdf.rb]: This method could really use some documentation.

File: lib/render_pdf.rb [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    body  # Add content, which is defined in subclasses
+    render  # Render pdf
+  end
+
+  # Helper method to test pdf via rails console: OrderByGroups.new(order).save_tmp
+  def save_tmp
+    File.open("#{Rails.root}/tmp/#{self.class.to_s.underscore}.pdf", 'w') {|f| f.write(to_pdf.force_encoding("UTF-8")) }
+  end
+
+  # XXX avoid underscore instead of unicode whitespace in pdf :/
```
Comment (wvengen) [inline lib/render_pdf.rb]: `@todo` is used in other places

File: lib/order_pdf.rb [hunk 5]
```diff
@@ -1,104 +1,130 @@
...
+  def each_order_article(&block)
+    OrderArticle.
+      ordered.
+      includes(article: :supplier).
+      includes(group_order_articles: {group_order: :ordergroup}).
+      where(order: @orders).
+      order('suppliers.name, articles.name, groups.name').
+      preload(:article_price).
+      each(&block)
+  end
```
Comment (wvengen) [inline lib/order_pdf.rb]: Can we split the scope and each? This can help plugins to add preloading for data to add to the pdf, for example.

**Comment by paroga** [comment]
screen shot

**Comment by paroga** [comment]
@wvengen ping

File: app/documents/order_by_articles.rb [hunk 6]
```diff
@@ -18,59 +12,30 @@ def title
...
+
+      order_article.group_order_articles.each do |goa|
```
Comment (wvengen) [inline app/documents/order_by_articles.rb]: Would be good to use a function getting the goa's, e.g. if a plugin wants to add an additional field hat needs to be selected from the database.

**Pull request merged**